otherlibs/dune-rpc/private/registry.ml

```diff
@@ -20,13 +20,11 @@ module Dune = struct
       ; where : Where.t
       }
 
-    let compare t { root; pid; where } : Stdune.Ordering.t =
-      match Int.compare t.pid pid with
-      | (Lt | Gt) as s -> s
-      | Eq -> (
-        match String.compare t.root root with
-        | (Lt | Gt) as s -> s
-        | Eq -> Where.compare t.where where)
+    let compare t { root; pid; where } : Ordering.t =
+      let open Ordering.O in
+      let= () = Int.compare t.pid pid in
+      let= () = String.compare t.root root in
+      Where.compare t.where where
```

Comment:
We should just use polymorphic comparison for simple cases like this.

Reply:
Note though that using Pervasives.compare on a record will make a C call every time while code like that quoted will only make a C call if the pids are equal. Maybe that matters if this is hot and the pid's are often different.

Reply:
Polymorphic comparisons are a time bomb. They continue working when the underlying data type changes in a way that makes the polymorphic comparison no longer acceptable. Perhaps, annotating every polymorphic comparison with a type helps catch such future changes but it seems better to just avoid polymorphic comparisons altogether.